### ISSUE TITLE
Should `shared` variables be included and used like global parameter by default?

### ISSUE THREAD

**Issue by gaow** [vatlab/SoS#1172]
One of my previous pipeline is broken. A MWE version of it is:

```
[1:shared='files']
files = ['1.txt', '2.txt', '3.txt', '4.txt']

[2]
sos_run('test',files=files)

[test]
print(files)
```

I understand i should add `parameter: files = list` to step `test`. But there are many parameters like this, and there are many steps using them. But more fundamentally, when something is `shared` should we expect it to be available in all "downstream" steps? ("downstream" is a tricky term but you get my point)

**Comment by BoPeng** [comment]
OK, this is #1164. 

So basically you want to use `files` from `sos_run`, not from `shared='files'`?

When something is `shared`, the step that produces `shared` will be depended upon by all steps that use this variable. In this case, `files` should either be `shared` and be provided by step `1`, or be parameter that is provided from command line or `sos_run`,  so I think there is some ambiguity here at step `test`.

**Comment by gaow** [comment]
> So basically you want to use files from sos_run, not from shared='files'?

Good question -- I dont know! Here I meant to use that from `sos_run` not from `shared` but I see how `files` is legit input anyways due to `shared` and we can possibly remove `files=files` from `sos_run`? But ambiguity is certainly not good. I cannot wrap my head around this one for now how to make it not ambiguous, though ...

**Comment by BoPeng** [comment]
So before `#1164`, 

```
sos_run('wf', a=1)
```
basically makes `a` available to all steps of `wf`, effectively creating something like

```
[global]
a=1

[wf]
```

Because of the availability of `a` in the global namespace, 

1. `parameter: a=1` will not be processed.
2. `shared='a'` will not be triggered because `a` already exists

#1164 enforces 1 but disallows 2. Whether 2 should be an allowed use case is up to discussion though.

**Comment by gaow** [comment]
> #1164 enforces 1

Just to clarify: it means we'll have to explicitly specify `parameter: a=1` or `parameter: a=int` either in `global` or `wf`, in order for `sos_run` to work like that, right?

> Whether 2 should be an allowed use case is up to discussion though.

The question is basically what is the scope of `shared` -- what is its current scope? In the context of process oriented workflow it makes sense to think `shared` only works for the same workflow, eg, `wf_*`. But of course this is just in my head. Thinking again, it sounds reasonable to have `shared` become basically read-only global parameters? That case we'd have this workflow: 


```
[1:shared='files']
files = ['1.txt', '2.txt', '3.txt', '4.txt']

[2]
sos_run('test')

[test]
print(files)
```

where step `test` does not make sense by itself unless one writes it as : 

```
[test]
depends: sos_variable(files)
print(files)
```

which is not required. And the `default` workflow is a bit weird since it seems to share some variables not used in the next workflow step -- but in fact it is used!

That said, making `shared` a read-only global variable does not sound a bad idea ...

**Comment by BoPeng** [comment]
> Just to clarify: it means we'll have to explicitly specify parameter: a=1 or parameter: a=int either in global or wf, in order for sos_run to work like that, right?

Yes, current there must be either a global or local `parameter: a`.

**Comment by BoPeng** [comment]
`shared` is designed to be more or less

```
[step: provides=sos_variable('var')]
```
so that it provides `var` when `var` is not defined. This is why `step` will not be called when `var` is defined in anyway (in your case through parameter of `sos_run`.

**Comment by gaow** [comment]
> so that it provides var when var is not defined.

Okay I can accept that behavior. But that means:

```
[1:shared='files']
files = ['1.txt', '2.txt', '3.txt', '4.txt']

[2]
sos_run('test')

[test]
print(files)              
``` 

should still work, since `files` is not defined otherwise, and "shared" `files` should kick in anyways? Currently the workflow above does not work.

**Comment by gaow** [comment]
This below, however, indeed works:

```
[1:shared='files']
files = ['1.txt', '2.txt', '3.txt', '4.txt']

[2]
sos_run('test')

[test]
depends: sos_variable('files')
print(files)
```

and I can accept that behavior as well, when I view `shared` as a shorthand for `provides=sos_variable('var')` ... it is just how we like to define the scope of `shared`.

**Comment by BoPeng** [comment]
SoS made an effort to make the following work but right now it can only detect dependency in few limited cases.

```
[1:shared='files']
files = ['1.txt', '2.txt', '3.txt', '4.txt']

[2]
sos_run('test')

[test]
print(files) 
```
The problem is that Python is very flexible and we have no clear way to know if `files` should be generated from another step. For example,

```
[test]
exec('a.py')
print(files)
```
could potentially make files available but there is no way SoS could detect it. So either

1. SoS gets all names in the script and calls steps with `shared='files'` even though `files` might be local to the step. or 
2. SoS requires explicit `depends: sos_variable('files')`.

2 seems to be a clearer way.

**Comment by gaow** [comment]
> to make it a true auxiliary step?

sure but that was not my point ... my main point is that I do not want it (the `files` step below) be execute multiple times

```
[files:shared='files']
files = ['1.txt', '2.txt', '3.txt', '4.txt']

[1]
sos_run('test')

[2]
sos_run('another_test')

[test]
depends: sos_variable('files')
print(files)

[another_test]
depends: sos_variable('files')
print(files)
```

**Comment by BoPeng** [comment]
> Even if I do default:2-3 I still run 1 twice ... how can I avoid running 1 multiple times?

This is DAGs of nested workflows are built separately from the main workflow. If you use

```
[A:shared='files']
files = ['1.txt', '2.txt', '3.txt', '4.txt']

[2]
depends: sos_step('test')

[3]
depends: sos_step('another_test')

[test]
depends: sos_variable('files')
print(files)

[another_test]
depends: sos_variable('files')
print(files)
```

Then `A` would be called only once.

**Comment by gaow** [comment]
sure, but I find it hard to think of for the forward logic in https://github.com/vatlab/jnbinder/blob/master/release#L257

Basically you are saying I should change the 3 `sos_run` calls to `depends`. But the `sos_run` calls are supposed to happen in a sequence. using `depends` kind of breaks the sequential logic, even if under the hood they are executed sequentially. 

Again I should be able to come up with a better design for that case but somehow it will end up outcome oriented design which makes me wonder if something simpler in logic can be done under existing framework.

**Comment by BoPeng** [comment]
No, I did not advertise the use of `sos_step`, which is backward thinking and  less intuitive than the forward-thinking `sos_run`. I am saying that `sos_run` builds and execute complete workflows so steps cannot be shared among them. Actually, because `sos_run()` sets a global namespace for the steps in the nested workflow, something like the following could be done and sharing steps would be wrong

```

[files: shared='files']
parameter: n=int
files = [f'file_{x}.txt' for x in range(n)]

[A]
depends: sos_variable('files')
print(files)

[B]
depends: sos_variable('files')
print(files)

[default]
sos_run('A', n=3)
sos_run('B', n=4)
```

**Comment by gaow** [comment]
Indeed. Still the issue is that when `n` is fixed the `files` step gets executed twice. 

What do you think about this pipeline though? 

https://github.com/vatlab/jnbinder/blob/master/release

what is the best way to re-write it? Notice the `default` is only 2 steps so it should be quite easy to follow the logic. If we see an elegant way out using existing framework that means our design withstands the challenge. Because current structure in that pipeline is quite straightforward I wrote it almost without designing it carefully and it worked.

**Comment by BoPeng** [comment]
Then something like

```
[default]
vars =  ('var1', 'var2', 'var3')
depends: [sos_variable(x) for x in vars]
sos_run('A', args=vars)
sos_run('B', args=vars)
```

`args=vars` does not work, but according to our convention, 

```
args='name'
```
can reasonably considered as
```
args=dict(name=name)
```

**Comment by gaow** [comment]
I ended up adding explicitly the parameters without rewriting the workflow:

https://github.com/vatlab/jnbinder/commit/5767acf8d2eca0018f54db89ee140e56c7b1a645

it is verbose but conceptually is more readable in a way. 

We can close this ticket if you like. #1164 has made the behavior more transparent though seems restricting in this particular example. But in the course of modifying those Process Oriented workflow examples using named outputs in combination with parameters, I think we've now got a more powerful system that makes `shared` less needed in the future.
